lti_consumer/plugin/views.py

```diff
@@ -231,3 +240,153 @@ def scores(self, request, *args, **kwargs):
         )
+
+
+class LtiNrpsViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    """
+    Api Endpoints for LTI NPRS Services
+    """
+    serializer_class = LtiNrpsContextMembershipSerializer
+
+    # Custom permission classes for LTI APIs
+    authentication_classes = [Lti1p3ApiAuthentication]
+    permission_classes = [LtiNrpsPermissions]
+
+    # Renderer classes to accept LTI NRPS content types
+    renderer_classes = [
+        MembershipResultRenderer,
+    ]
+
+    def get_page_size(self):
+        """
+        Get Page size for pagination
+        """
+        return settings.REST_FRAMEWORK.get('PAGE_SIZE', 10)
+
+    def build_queryset(self, filter_by=None):
+        """
+        Given a course key return a list of enrollments.
+        """
+        # pylint: disable=import-error,import-outside-toplevel,no-name-in-module
+        from common.djangoapps.student.models import CourseAccessRole
+
+        course_key = self.request.lti_configuration.location.course_key
+
+        User = get_user_model()
+
+        # prepare prefetch query for course access role to speed up lookup
+        pref_course_access_role = Prefetch(
+            'courseaccessrole_set',
+            CourseAccessRole.objects.filter(
+                course_id=course_key, role__in=['instructor', 'staff'])
+        )
+
+        # fetch users that has either an access role or an enrollment
+        queryset = User.objects.filter()
+
+        if filter_by:
+            if filter_by == 'student':
+                queryset = queryset.filter(
+                    courseenrollment__course_id=course_key,
+                    courseenrollment__is_active=True
+                )
+            elif filter_by == 'instructor':
+                queryset = queryset.filter(
+                    courseaccessrole__course_id=course_key,
+                    courseaccessrole__role='instructor'
+                )
+            elif filter_by == 'staff':
+                queryset = queryset.filter(
+                    courseaccessrole__course_id=course_key,
+                    courseaccessrole__role='staff'
+                )
+        else:
+            queryset = queryset.filter(
+                Q(courseenrollment__course_id=course_key, courseenrollment__is_active=True) |
+                Q(courseaccessrole__course_id=course_key)
+            )
+
+        # distinct prevents duplicates when matching on many to many relation
+        queryset = queryset.prefetch_related(pref_course_access_role).distinct()
+
+        return queryset
+
+    @action(methods=['GET'], detail=False, url_path='memberships')
+    def memberships(self, request, *args, **kwargs):
+        """
+        LTI NRPS Context Membership service endpoint.
+
+        See full API spec on -
+        https://www.imsglobal.org/spec/lti-nrps/v2p0#context-membership
+        """
+        role_filter_param = self.request.query_params.get('role', None)
+
+        # get current page number, make sure it's an integer
```

Comment:
Instead of building the paginator mechanism like this, can we use a custom paginator class? 
Ref: https://www.django-rest-framework.org/api-guide/pagination/#custom-pagination-styles

Reply:
Yeah, added a custom paginator class and removed these portion.